pkg/scm/git/git.go

```diff
@@ -62,18 +69,145 @@ func stringInSlice(s string, slice []string) bool {
 // ValidCloneSpec determines if the given string reference points to a valid git
 // repository
 func (h *stiGit) ValidCloneSpec(source string) bool {
-	url, err := url.Parse(source)
-	if err != nil {
-		return false
+	rc, _, _, _, _, _ := validCloneSpec(source, false, nil)
+	return rc
+}
+
+// ValidCloneSpecRemoteOnly determines if the given string reference points to a valid remote git
+// repository; valid local repository specs will result in a return of false
+func (h *stiGit) ValidCloneSpecRemoteOnly(source string) bool {
+	rc, _, _, _, _, _ := validCloneSpec(source, true, nil)
+	return rc
+}
+
+// MungeNoProtocolURL will take a URL returned from net/url.Parse and make
+// corrections to the URL when no protocol is specified in the Scheme, where there
+// are valid protocol-less git url spec formats that result in either file:// or ssh:// protocol usage;
+// if an explicit protocol other than file:// or ssh:// are already specified in a valid way, then the
+// URL is left unchaged and the method simply returns with no error reported,
+// since the URL is
+func (h *stiGit) MungeNoProtocolURL(source string, uri *url.URL) error {
+	if uri == nil {
+		return nil
+	}
+	rc, scheme, user, host, path, ref := validCloneSpec(source, false, uri)
+	if !rc {
+		return errors.New(fmt.Sprintf("invalid git clone spec provided, source string %s and URL %v", source, uri))
+	}
+	// don't modify the protocol options which always have to be explicitly specified (http, https,  git)
+	if scheme != "ssh" && scheme != "file" {
+		return nil
+	}
+
+	// update the either file or ssh url accordingly
+	if len(user) > 0 {
+		uri.User = url.User(user)
+	}
+	uri.Scheme = scheme
+	uri.Host = host
+	uri.Path = path
+	uri.Fragment = ref
+	return nil
+}
+
+func schemeNecessitatesReturn(uri *url.URL) (shouldReturn, returnCode bool) {
```

Comment:
It's not clear from the name of the function what it does or what the return values mean. At the very least you should have a comment describing what it does.